### PR TITLE
Match

### PR DESCRIPTION
**Pull request by alradish**

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/intentions/AddMatchWildcardPatternIntention.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RsBlockExpr
+import org.rust.lang.core.psi.RsMatchExpr
+import org.rust.lang.core.psi.RsPatWild
+import org.rust.lang.core.psi.RsPsiFactory
+import org.rust.lang.core.psi.ext.ancestorStrict
+
+class AddMatchWildcardPatternIntention : RsElementBaseIntentionAction<AddMatchWildcardPatternIntention.Context>() {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/intentions/AddMatchWildcardPatternIntention.kt]: Split it into 2 separate PR. Pattern matching is a part of language support story, this intention is not directly related to it.

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+fun RsPath.singleOrVariant(pat: RsPat): List<Constructor> {
+    val item = reference.resolve()
+    return if (item is RsEnumVariant) listOf(Variant(pat, item))
+    else listOf(Single(pat))
+}
+
+
+val List<List<*>>.width: Int
+    get() {
+        return this.maxWith(Comparator { p0, p1 ->
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Comparator.comparing... ?

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+              if slice.is_some() {
+                  Some((pat_len..pcx.max_slice_length+1).map(Slice).collect())
+              } else {
+                  Some(vec![Slice(pat_len)])
+              }
+        }
+    }
+}*/
+
+val RsPat.constructors: List<Constructor>
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: please, give to this function more meaningful name

**Comment by sirgl** [comment]
In general ok, but seems, that it would be hard to achieve our goals soon. We should concentrate on full cycle of inspection + fix work, not only recognising. If language support is incomplete, but in some cases it works, it would be more helpful, than if LS would be complete, but PR won't be accepted due to absence of uses.

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+        val isPrivatelyEmpty = allConstructors.isEmpty()
+        val isDeclaredNonexhaustive = (ty as? TyAdt)?.isNonExhaustiveEnum ?: false
+        println("<top>.isUseful missing constructors $missingConstructor\tis privately empty $isPrivatelyEmpty, is declared nonexhaustive $isDeclaredNonexhaustive")
+
+        val isNonExhaustive = isPrivatelyEmpty || isDeclaredNonexhaustive
+
+        return if (missingConstructor.isEmpty() && !isNonExhaustive) {
+            allConstructors.map {
+                isUsefulS(matrix, v, it)
+            }.find {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Isn't it alConstructors.any { ... } ?

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+            if (pair.second == null) {
+                seen.add(v)
+                if (catchAll == false && false*//*check catchall for RsPat*//*) catchAll = true
+            }
+        }
+    }
+}*/
+
+// Use algorithm from 3.1 http://moscova.inria.fr/~maranget/papers/warn/warn004.html
+fun isUseful(matrix: List<List<Pattern>?>, v: List<Pattern>): Boolean {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Is it really required to have `?` here ? Can't we just replace with empty list?

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+            }.mapNotNull { it }
+            isUseful(newMatrix, v.subList(1, v.size))
+        }
+    }
+}
+
+fun isUsefulS(matrix: List<List<Pattern>?>, v: List<Pattern>, constructor: Constructor): Boolean {
+    println("<top>.isUsefulS(matrix = $matrix, v = $v, constructor = $constructor)")
+
+    val newMatrix = matrix.map { specializeRow(it, constructor) }.mapNotNull { it }
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Same for this place.

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+
+    //// Induction
+
+
+    val constructors = v[0].constructors
+
+    println("<top>.isUseful expand first col: expanding ${v[0]}")
+    return if (constructors != null) {
+        println("<top>.isUseful expanding constructors $constructors")
+        constructors.map {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Duplicates with 127-131, should be refactored

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+        val isNonExhaustive = isPrivatelyEmpty || isDeclaredNonexhaustive
+
+        return if (missingConstructor.isEmpty() && !isNonExhaustive) {
+            allConstructors.map {
+                isUsefulS(matrix, v, it)
+            }.find {
+                it
+            } ?: false
+        } else {
+            val newMatrix = matrix.map {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Just use `mapNotNull`, it will return only not null values.

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+        println("<top>.isUseful expanding constructors $constructors")
+        constructors.map {
+            isUsefulS(matrix, v, it)
+        }.find { it == true } ?: false
+    } else {
+        println("<top>.isUseful expanding wildcard")
+        val usedConstructors = matrix.map { it?.get(0)?.constructors }.flatMap { it ?: emptyList() }
+        println("<top>.isUseful used constructors $usedConstructors")
+
+        val ty = matrix.mapNotNull { it?.get(0) }
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Why do you chain map's? It creates list on every call. Just extract `ty` inside single lambda

Comment (sirgl) [reply src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: Also, here `map` is not even need. You can use `matrix.find { it?.get(0)?.ty != TyUnknown } `
Probably you should extract function inside and give it meaningful name.

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+    get() = matchBody?.matchArmList?.map { arm ->
+        arm.patList.map { lowerPattern(it) } to arm.matchArmGuard
+    } ?: emptyList()
+
+
+val TyAdt.isNonExhaustiveEnum: Boolean
+    get() {
+        val enum = item as? RsEnumItem ?: return false
+        val attrList = enum.outerAttrList
+        val attr = attrList.find {
```
Comment (sirgl) [inline src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: return attrList.any { it.metaItem.name == "non_exhaustive" }

File: src/main/kotlin/org/rust/lang/core/psi/RsLiteralKind.kt [hunk 11]
```diff
@@ -35,6 +35,17 @@ sealed class RsLiteralKind(val node: ASTNode) {
         override val validSuffixes: List<kotlin.String>
             get() = listOf("u8", "i8", "u16", "i16", "u32", "i32", "u64", "i64", "u128", "i128", "isize", "usize")
 
+        val value: Int?
```
Comment (sirgl) [inline src/main/kotlin/org/rust/lang/core/psi/RsLiteralKind.kt]: `Int` is not capable to store `i132`. Let's use Number here (`NumberFormat.getInstance().parse(text)`)

File: src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt (hunk 8 above)
Comment (alradish) [reply src/main/kotlin/org/rust/ide/inspections/RsMatchCheckInspection.kt]: thx

**Comment by alradish** [comment]
some
